### PR TITLE
`Integrated code lifecycle`: Implement a workaround for hanging build jobs

### PR DESCRIPTION
**Pull request by BBesrour**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).

### Motivation and Context

Fix hanging build jobs due threadpool being full. Reason why buildagent is marked as available even though pool is full is still unknown

### Steps for Testing

Server with ICL
Prerequisites:
- 1 Student
- 1 Programming Exercise

1. Check if programming exercises work correctly

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [ ] Test 1
- [x] Test 2

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved reliability of build job processing by handling `RejectedExecutionException`. If an exception occurs, the job is now re-added to the queue, and local build agent information is updated.

### CODE REVIEW

**Review by az108 — APPROVED** [review]
Tested on TS3, programming exercises work as expected

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/buildagent/SharedQueueProcessingService.java [hunk 1]
```diff
@@ -184,6 +183,22 @@ private void checkAvailabilityAndProcessNextBuild() {
             }
             processBuild(buildJob);
         }
+        catch (RejectedExecutionException e) {
+            log.error("Couldn't add build job to threadpool: {}\n Concurrent Build Jobs Count: {} Active tasks in pool: {}, Concurrent Build Jobs Size {}", buildJob,
```
Comment (Jan-Thurner) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localci/buildagent/SharedQueueProcessingService.java:187]:
```suggestion
            log.error("Couldn't add build job to threadpool: {}\n Concurrent Build Jobs Count: {} Active tasks in pool: {}, Concurrent Build Jobs Size: {}", buildJob,
```

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/buildagent/SharedQueueProcessingService.java [hunk 2]
```diff
@@ -347,8 +362,8 @@ private void processBuild(BuildJobQueueItem buildJob) {
      * Checks whether the node has at least one thread available for a new build job.
      */
     private boolean nodeIsAvailable() {
-        log.debug("Currently processing jobs on this node: {}, maximum pool size of thread executor : {}", localProcessingJobs.get(),
-                localCIBuildExecutorService.getMaximumPoolSize());
+        log.debug("Currently processing jobs on this node: {}, active threads in Pool {}, maximum pool size of thread executor : {}", localProcessingJobs.get(),
```
Comment (Jan-Thurner) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localci/buildagent/SharedQueueProcessingService.java:365]:
```suggestion
        log.debug("Currently processing jobs on this node: {}, active threads in Pool: {}, maximum pool size of thread executor : {}", localProcessingJobs.get(),
```

**Review by Jan-Thurner — APPROVED** [review]
Tested on TS3, works as expected. Code also lgtm, but you might consider adding `:` to the additional error messages (see comments). However, I do not have a strong opinion on this.

**Review by EneaGore — APPROVED** [review]
Tested on TS5. Programming Exercise works as expected.

**Comment by BBesrour** [comment]
@EneaGore @undernagruzez Thanks for your review! However, TS5 doesn't use ICL. Please use TS1, 2, 3 or 4.

**Review by az108 — APPROVED** [review]
Change LGTM, Reapproval

**Review by Jan-Thurner — APPROVED** [review]
Changes LGTM

**Review by JohannesStoehr — APPROVED** [review]
Code

**Pull request merged**